docs/api/api-config.md
<!-- ***********************************************************************************************
	Config API
************************************************************************************************ -->
# `Config` API {#config-api}

The `Config` object controls various aspects of SugarCube's behavior.

<p role="note"><b>Note:</b>
<code>Config</code> object settings should be placed within your project's JavaScript section (Twine&nbsp;2: the Story JavaScript; Twine&nbsp;1/Twee: a <code>script</code>-tagged passage).
</p>


<!-- ***************************************************************************
	General
**************************************************************************** -->
## General Settings<!-- legacy --><span id="config-api-miscellaneous"></span><!-- /legacy --> {#config-api-general}

<!-- *********************************************************************** -->

### `Config.addVisitedLinkClass` ↔ `boolean` (default: `false`) {#config-api-property-addvisitedlinkclass}

Determines whether the `link-visited` class is added to internal passage links that go to previously visited passages—i.e., the passage already exists within the story history.

<p role="note"><b>Note:</b>
You <em>must</em> provide your own styling for the <code>link-visited</code> class as none is provided by default.
</p>

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `boolean` value signifying whether links to visited passages have the class added.

#### Examples:

```javascript
Config.addVisitedLinkClass = true;
```

#### CSS styles:

You will also need to specify a `.link-visited` style that defines the properties visited links should have. For example:

```css
.link-visited {
	color: purple;
}
```

<!-- *********************************************************************** -->

### `Config.cleanupWikifierOutput` ↔ `boolean` (default: `false`) {#config-api-property-cleanupwikifieroutput}

Determines whether the output of the Wikifier is post-processed into more sane markup—i.e., where appropriate, it tries to transition the plethora of `<br>` elements into `<p>` elements.

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `boolean` value signifying whether post-processing occurs.

#### Examples:

```javascript
Config.cleanupWikifierOutput = true;
```

<!-- *********************************************************************** -->

### `Config.debug` ↔ `boolean` (default: `false`) {#config-api-property-debug}

Indicates whether SugarCube is running in test mode, which enables debug views and various optional debugging errors and warnings.  See the [*Test Mode* guide](#guide-test-mode) for more information.

<p role="note"><b>Note:</b>
This setting is automatically set based on whether you're using a testing mode in a Twine compiler—i.e., <em>Test</em> mode in Twine&nbsp;2, <em>Test Play From Here</em> in Twine&nbsp;1, or the test mode option (<code>-t</code>, <code>--test</code>) in Tweego.  You may, however, forcibly enable it if you need to for some reason—e.g., if you're using another compiler, which doesn't offer a way to enable test mode.
</p>

<p role="note" class="see"><b>See Also:</b>
<a href="#config-api-property-enableoptionaldebugging"><code>Config.enableOptionalDebugging</code> setting</a>.
</p>

#### History:

* `v2.2.0`: Introduced.

#### Value:

A `boolean` value signifying whether SugarCube is running in test mode.

#### Examples:

##### Forcibly enabling test mode

```javascript
// Forcibly enable test mode.
Config.debug = true;
```

##### Check if test mode is enabled (in JavaScript)

```javascript
if (Config.debug) {
	// Do something debug related.
}
```

##### Check if test mode is enabled (in macros)

```
<<if Config.debug>>
	/* do something debug related */
<</if>>
```

<!-- *********************************************************************** -->

### `Config.enableOptionalDebugging` ↔ `boolean` (default: `false`) {#config-api-property-enableoptionaldebugging}

Determines whether various optional debugging errors and warnings are enabled outside of test mode.

<p role="note" class="see"><b>See Also:</b>
<a href="#config-api-property-debug"><code>Config.debug</code> setting</a>.
</p>

List of optional errors and warnings: *(not exhaustive)*

* [`<<if>>` macro](#macros-macro-if) assignment error.  If enabled, returns an error when the `=` assignment operator is used within its conditional—e.g., `<<if $suspect = "Bob">>`.  Does not flag other assignment operators.

#### History:

* `v2.37.0`: Introduced.

#### Value:

A `boolean` value signifying whether optional debugging errors and warnings are disabled.

#### Examples:

```javascript
Config.enableOptionalDebugging = true;
```

<!-- *********************************************************************** -->

### `Config.loadDelay` ↔ *integer* `number` (default: `0`) {#config-api-property-loaddelay}

Sets the integer delay (in milliseconds) before the loading screen is dismissed, once the document has signaled its readiness.  Not generally necessary, however, some browsers render slower than others and may need a little extra time to get a media-heavy page done.  This allows you to fine tune for those cases.

#### History:

* `v2.0.0`: Introduced.

#### Value:

An *integer* `number` value denoting the delay (in milliseconds) before the loading screen is dismissed.

#### Examples:

```javascript
// Delay the dismissal of the loading screen by 2000ms (2s)
Config.loadDelay = 2000;
```


<!-- ***************************************************************************
	Audio
**************************************************************************** -->
## Audio Settings {#config-api-audio}

<!-- *********************************************************************** -->

### `Config.audio.pauseOnFadeToZero` ↔ `boolean` (default: `true`) {#config-api-property-audio-pauseonfadetozero}

Determines whether the audio subsystem automatically pauses tracks that have been faded to `0` volume (silent).

#### History:

* `v2.28.0`: Introduced.

#### Value:

A `boolean` value signifying whether tracks that have faded to `0` are automatically paused.

#### Examples:

```javascript
Config.audio.pauseOnFadeToZero = false;
```

<!-- *********************************************************************** -->

### `Config.audio.preloadMetadata` ↔ `boolean` (default: `true`) {#config-api-property-audio-preloadmetadata}

Determines whether the audio subsystem attempts to preload track metadata—meaning information about the track (e.g., duration), not its audio frames.

<p role="note"><b>Note:</b>
It is unlikely that you will ever want to disable this setting.
</p>

#### History:

* `v2.28.0`: Introduced.

#### Value:

A `boolean` value signifying whether track metadata is preloaded.

#### Examples:

```javascript
Config.audio.preloadMetadata = false;
```


<!-- ***************************************************************************
	History
**************************************************************************** -->
## History Settings {#config-api-history}

<!-- *********************************************************************** -->

### `Config.history.controls` ↔ `boolean` (default: `true`) {#config-api-property-history-controls}

Determines whether the history controls (Backward, Jump To, & Forward buttons) are enabled within the UI bar.

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `boolean` value signifying whether the history controls are enabled.

#### Examples:

```javascript
Config.history.controls = false;
```

<!-- *********************************************************************** -->

### `Config.history.disableDeltas` ↔ `boolean` (default: `false`) {#config-api-property-history-disabledeltas}

Determines whether the history states are delta encoded/decoded.

<p role="note" class="tip"><b>Tip:</b>
For the <em>vast majority</em> of projects it is <strong><em>strongly recommended</em></strong> to leave this setting at the default value of <code>false</code>.
</p>

<p role="note" class="warning"><b>Warning:</b>
Changing this setting's value will invalidate all saves and sessions generated while the opposite value was in effect.
</p>

<p role="note" class="warning"><b>Warning:</b>
Private fields in classes are only supported when this setting is set to <code>true</code>.
</p>

#### History:

* `v2.38.0`: Introduced.

#### Value:

A `boolean` value signifying whether delta encoding/decoding of the history states is disabled.

#### Examples:

```javascript
// Disables delta encoding/decoding of the history.
Config.history.disableDeltas = true;
```

<!-- *********************************************************************** -->

### `Config.history.maxStates` ↔ *integer* `number` (default: `40`) {#config-api-property-history-maxstates}

Sets the maximum number of states (moments) to which the history is allowed to grow.  Should the history exceed the limit, states will be dropped from the past (oldest first).

<p role="note" class="tip"><b>Tip:</b>
For game-oriented projects, as opposed to more story-oriented interactive fiction, a setting of <code>1</code> is <strong><em>strongly recommended</em></strong>.
</p>

#### History:

* `v2.0.0`: Introduced.
* `v2.36.0`: Reduced the default to `40`.

#### Value:

An *integer* `number` value denoting how many moments are allowed within the history.

#### Examples:

```javascript
// Limit the history to a single moment (recommended for games)
Config.history.maxStates = 1;

// Limit the history to 25 moments
Config.history.maxStates = 25;
```


<!-- ***************************************************************************
	Macros
**************************************************************************** -->
## Macros Settings {#config-api-macros}

<!-- *********************************************************************** -->

### `Config.macros.maxLoopIterations` ↔ *integer* `number` (default: `1000`) {#config-api-property-macros-maxloopiterations}

Sets the maximum number of iterations allowed before the [`<<for>>` macro](#macros-macro-for) conditional forms are terminated with an error.

<p role="note"><b>Note:</b>
This setting exists to prevent a misconfigured loop from making the browser unresponsive.
</p>

#### History:

* `v2.0.0`: Introduced.

#### Value:

An *integer* `number` value denoting the total allowed number of conditional form loop iteration (per instance).

#### Examples:

```javascript
// Allow only 5000 iterations
Config.macros.maxLoopIterations = 5000;
```

<!-- *********************************************************************** -->

### `Config.macros.typeSkipKey` ↔ `string` (default: `" "`, space) {#config-api-property-macros-typeskipkey}

Sets the default [`KeyboardEvent.key`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key) value that causes the currently running [`<<type>>` macro](#macros-macro-type) instance to finish typing its content immediately.

#### History:

* `v2.33.1`: Introduced.

#### Value:

A `string` value denoting the keyboard key that causes `<<type>>` to immediately finish typing.

#### Examples:

```javascript
// Change the default skip key to Control (CTRL)
Config.macros.typeSkipKey = 'Control';
```

<!-- *********************************************************************** -->

### `Config.macros.typeVisitedPassages` ↔ `boolean` (default: `true`) {#config-api-property-macros-typevisitedpassages}

Determines whether the [`<<type>>` macro](#macros-macro-type) types out content on previously visited passages or simply outputs it immediately.

#### History:

* `v2.32.0`: Introduced.

#### Value:

A `boolean` value signifying whether `<<type>>` types out previously visited passages.

#### Examples:

```javascript
// Do not type on previously visited passages
Config.macros.typeVisitedPassages = false;
```

<!-- *********************************************************************** -->

### <span class="deprecated">`Config.macros.ifAssignmentError` ↔ `boolean` (default: `true`)</span> {#config-api-property-macros-ifassignmenterror}

<p role="note" class="warning"><b>Deprecated:</b>
This setting has been deprecated and should no longer be used.  See the <a href="#config-api-property-enableoptionaldebugging"><code>Config.enableOptionalDebugging</code></a> setting for its replacement.
</p>

#### History:

* `v2.0.0`: Introduced.
* `v2.37.0`: Deprecated in favor of the `Config.enableOptionalDebugging` setting.


<!-- ***************************************************************************
	Navigation
**************************************************************************** -->
## Navigation Settings {#config-api-navigation}

<!-- *********************************************************************** -->

### `Config.navigation.override` ↔ `Function` (default: *none*) {#config-api-property-navigation-override}

Allows the destination of passage navigation to be overridden.  The callback is passed one parameter, the original destination passage name.  If its return value is falsy, the override is cancelled and navigation to the original destination continues unperturbed.  If its return value is truthy, the override succeeds and that value is used as the new destination of the navigation.

#### History:

* `v2.13.0`: Introduced.

#### Value:

The callback value (`Function`).

#### Callback parameters:

* **`destinationPassage`:** (`string`) The original destination passage name.

#### Examples:

##### Basic usage

```javascript
Config.navigation.override = (destinationPassage) => {
	/* code that returns a passage name or a falsy value */
};
```

##### Based upon a story variable

```javascript
Config.navigation.override = (destinationPassage) => {
	var sv = State.variables;

	// If $health is less-than-or-equal-to 0, go to the "You Died" passage instead.
	if (sv.health <= 0) {
		return 'You Died';
	}
};
```


<!-- ***************************************************************************
	Passages
**************************************************************************** -->
## Passages Settings {#config-api-passages}

<!-- *********************************************************************** -->

### `Config.passages.displayTitles` ↔ `boolean` (default: `false`) {#config-api-property-passages-displaytitles}

Determines whether the current passage name is combined with the story's name within the browser tab's title bar.

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `boolean` value signifying whether passage names are combined with the story's name within the title bar.

#### Examples:

```javascript
Config.passages.displayTitles = true;
```

<!-- *********************************************************************** -->

### `Config.passages.nobr` ↔ `boolean` (default: `false`) {#config-api-property-passages-nobr}

Determines whether rendering passages have their leading/trailing newlines removed and all remaining sequences of newlines replaced with single spaces before they're rendered.  Equivalent to including the [`nobr` special tag](#special-tag-nobr) on every passage.

<p role="note"><b>Note:</b>
Does not affect <code>script</code> or <code>stylesheet</code> tagged passages, for Twine&nbsp;1/Twee, or the Story JavaScript or Story Stylesheet sections, for Twine&nbsp;2.
</p>

#### History:

* `v2.19.0`: Introduced.

#### Value:

A `boolean` value signifying whether rendering passages have a global no-break enabled.

#### Examples:

```javascript
Config.passages.nobr = true;
```

<!-- *********************************************************************** -->

### `Config.passages.onProcess` ↔ `Function` (default: *none*) {#config-api-property-passages-onprocess}

Allows custom processing of passage text.  The function is invoked each time the [`<Passage>.processText()` method](#passage-api-prototype-method-processtext) is called.  It is passed an abbreviated version of the associated passage's [`Passage` instance](#passage-api)—containing only the `name`, `tags`, and `text` properties.  Its return value should be the post-processed text.

<p role="note"><b>Note:</b>
Does not affect <code>script</code> or <code>stylesheet</code> tagged passages, for Twine&nbsp;1/Twee, or the Story JavaScript or Story Stylesheet sections, for Twine&nbsp;2.
</p>

<div role="note"><b>Note:</b>
	<p>The function will be called just before the built-in no-break passage processing if you're also using that.</p>
	<p>See the <a href="#config-api-property-passages-nobr"><code>Config.passages.nobr</code> setting</a> and <a href="#special-tag-nobr"><code>nobr</code> special tag</a>.</p>
</div>

#### History:

* `v2.30.0`: Introduced.

#### Value:

A callback value (`Function`).

#### Callback parameters:

* **`passage`:** (`Object`) An abbreviated version of the associated passage's [`Passage` instance](#passage-api)—containing only the `name`, `tags`, and `text` properties.

#### Examples:

```javascript
// Change instances of "cat" to "dog" in passages
Config.passages.onProcess = (p) => p.text.replace(/\bcat(s?)\b/g, 'dog$1');
```


<!-- *********************************************************************** -->

### `Config.passages.start` ↔ `string` (Twine&nbsp;2 default: *user-selected*; Twine&nbsp;1/Twee default: `"Start"`) {#config-api-property-passages-start}

Sets the starting passage name, the very first passage that will be displayed.

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `string` value denoting the starting passage name.

#### Examples:

```javascript
Config.passages.start = 'That Other Starting Passage';
```

<!-- *********************************************************************** -->

### `Config.passages.transitionOut` ↔ `string` | *integer* `number` (default: *none*) {#config-api-property-passages-transitionout}

Determines whether outgoing passage transitions are enabled.  Valid values are the name of the property being animated, which causes the outgoing passage element to be removed once that transition animation is complete, or an integer delay (in milliseconds), which causes the outgoing passage element to be removed once the delay has expired.  You will also need some CSS styles to make this work—examples given below.

<p role="note"><b>Note:</b>
If using an integer delay, ideally, it should probably be slightly longer than the outgoing transition delay that you intend to use—e.g., an additional 25ms or so should be sufficient.
</p>

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `string` value denoting the CSS property being animated or an *integer* `number` value denoting the integer delay.

#### Examples:

```javascript
// Remove outgoing elements when their opacity animation ends

src/config.js
/***********************************************************************************************************************

	config.js

	Copyright © 2013–2025 Thomas Michael Edwards <thomasmedwards@gmail.com>. All rights reserved.
	Use of this source code is governed by a BSD 2-clause "Simplified" License, which may be found in the LICENSE file.

***********************************************************************************************************************/
/* global Save, State, Story, getTypeOf, warnDeprecated */

var Config = (() => { // eslint-disable-line no-unused-vars, no-var
	// General settings.
	let cfgAddVisitedLinkClass     = false;
	let cfgCleanupWikifierOutput   = false;
	let cfgDebug                   = false;
	let cfgEnableOptionalDebugging = false;
	let cfgLoadDelay               = 0;

	// Audio settings.
	let cfgAudioPauseOnFadeToZero = true;
	let cfgAudioPreloadMetadata   = true;

	// State history settings.
	let cfgHistoryControls      = true;
	let cfgHistoryDisableDeltas = false;
	let cfgHistoryMaxStates     = 40;

	// Macros settings.
	let cfgMacrosMaxLoopIterations   = 1000;
	let cfgMacrosTypeSkipKey         = '\x20'; // Space
	let cfgMacrosTypeVisitedPassages = true;

	// Navigation settings.
	let cfgNavigationOverride;

	// Passages settings.
	let cfgPassagesDisplayTitles = false;
	let cfgPassagesNobr          = false;
	let cfgPassagesStart; // NOTE: Initially set by `Story.init()`
	let cfgPassagesOnProcess;
	let cfgPassagesTransitionOut;

	// Saves settings.
	let cfgSavesDescriptions;
	let cfgSavesId; // NOTE: Initially set by `Story.init()`
	let cfgSavesIsAllowed;
	let cfgSavesEvalEnabled  = true;
	let cfgSavesMaxAuto      = 0;
	let cfgSavesMaxSlot      = 8;
	let cfgSavesMetadata;
	let cfgSavesVersion;
	/* [DEPRECATED] */
	let cfgSavesAutoload;
	/* [/DEPRECATED] */

	// UI settings.
	let cfgUiStowBarInitially    = 800;
	let cfgUiUpdateStoryElements = true;


	/*******************************************************************************
		Object Exports.
	*******************************************************************************/

	return Object.freeze({
		/*
			General settings.
		*/
		get addVisitedLinkClass() { return cfgAddVisitedLinkClass; },
		set addVisitedLinkClass(value) { cfgAddVisitedLinkClass = Boolean(value); },

		get cleanupWikifierOutput() { return cfgCleanupWikifierOutput; },
		set cleanupWikifierOutput(value) { cfgCleanupWikifierOutput = Boolean(value); },

		get debug() { return cfgDebug; },
		set debug(value) { cfgDebug = Boolean(value); },

		get enableOptionalDebugging() { return cfgEnableOptionalDebugging; },
		set enableOptionalDebugging(value) { cfgEnableOptionalDebugging = Boolean(value); },

		get loadDelay() { return cfgLoadDelay; },
		set loadDelay(value) {
			if (!Number.isSafeInteger(value) || value < 0) {
				throw new RangeError('Config.loadDelay must be a non-negative integer number');
			}

			cfgLoadDelay = value;
		},

		/*
			Audio settings.
		*/
		audio : Object.freeze({
			get pauseOnFadeToZero() { return cfgAudioPauseOnFadeToZero; },
			set pauseOnFadeToZero(value) { cfgAudioPauseOnFadeToZero = Boolean(value); },

			get preloadMetadata() { return cfgAudioPreloadMetadata; },
			set preloadMetadata(value) { cfgAudioPreloadMetadata = Boolean(value); }
		}),

		/*
			State history settings.
		*/
		history : Object.freeze({
			// TODO: (v3) This should be under UI settings → `Config.ui.historyControls`.
			get controls() { return cfgHistoryControls; },
			set controls(value) {
				const controls = Boolean(value);

				if (cfgHistoryMaxStates === 1 && controls) {
					throw new Error('Config.history.controls must be false when Config.history.maxStates is 1');
				}

				cfgHistoryControls = controls;
			},

			get disableDeltas() { return cfgHistoryDisableDeltas; },
			set disableDeltas(value) { cfgHistoryDisableDeltas = Boolean(value); },

			get maxStates() { return cfgHistoryMaxStates; },
			set maxStates(value) {
				if (!Number.isSafeInteger(value) || value < 1) {
					throw new RangeError('Config.history.maxStates must be a positive integer number');
				}

				cfgHistoryMaxStates = value;

				// Force `Config.history.controls` to `false`, when limited to `1` moment.
				if (cfgHistoryControls && value === 1) {
					cfgHistoryControls = false;
				}
			}
		}),

		/*
			Macros settings.
		*/
		macros : Object.freeze({
			get maxLoopIterations() { return cfgMacrosMaxLoopIterations; },
			set maxLoopIterations(value) {
				if (!Number.isSafeInteger(value) || value < 1) {
					throw new RangeError('Config.macros.maxLoopIterations must be a positive integer number');
				}

				cfgMacrosMaxLoopIterations = value;
			},

			get typeSkipKey() { return cfgMacrosTypeSkipKey; },
			set typeSkipKey(value) { cfgMacrosTypeSkipKey = String(value); },

			get typeVisitedPassages() { return cfgMacrosTypeVisitedPassages; },
			set typeVisitedPassages(value) { cfgMacrosTypeVisitedPassages = Boolean(value); }

			/* [DEPRECATED] */
			/* eslint-disable comma-style */
			// Die if the deprecated macros if assignment error getter is accessed.
			, get ifAssignmentError() {
				warnDeprecated(
					'Config.macros.ifAssignmentError',
					'Config.enableOptionalDebugging',
					true
				);
			}
			// Warn if the deprecated macros if assignment error setter is assigned to,
			// while also setting `Config.enableOptionalDebugging` for compatibilities sake.
			, set ifAssignmentError(value) {
				warnDeprecated(
					'Config.macros.ifAssignmentError',
					'Config.enableOptionalDebugging'
				);
				Config.enableOptionalDebugging = value;
			}
			/* eslint-enable comma-style */
			/* [/DEPRECATED] */
		}),

		/*
			Navigation settings.
		*/
		navigation : Object.freeze({
			get override() { return cfgNavigationOverride; },
			set override(value) {
				if (!(value == null || value instanceof Function)) { // nullish test
					throw new TypeError(`Config.navigation.override must be a function, null, or undefined (received: ${getTypeOf(value)})`);
				}

				cfgNavigationOverride = value;
			}
		}),

		/*
			Passages settings.
		*/
		passages : Object.freeze({
			// TODO: (v3) This should be under Navigation settings → `Config.navigation.updateTitle`.
			get displayTitles() { return cfgPassagesDisplayTitles; },
			set displayTitles(value) { cfgPassagesDisplayTitles = Boolean(value); },

			get nobr() { return cfgPassagesNobr; },
			set nobr(value) { cfgPassagesNobr = Boolean(value); },

			get onProcess() { return cfgPassagesOnProcess; },
			set onProcess(value) {
				if (value != null) { // nullish test
					const valueType = getTypeOf(value);

					if (valueType !== 'function') {
						throw new TypeError(`Config.passages.onProcess must be a function, null, or undefined (received: ${valueType})`);
					}
				}

				cfgPassagesOnProcess = value;
			},

			// TODO: (v3) This should be under Navigation settings → `Config.navigation.startPassageName`.
			get start() { return cfgPassagesStart; },
			set start(value) {
				if (value != null) { // nullish test
					const valueType = getTypeOf(value);

					if (valueType !== 'string') {
						throw new TypeError(`Config.passages.start must be a string, null, or undefined (received: ${valueType})`);
					}
				}

				cfgPassagesStart = value;
			},

			// TODO: (v3) This should be under Navigation settings → `Config.navigation.transitionOut`.
			get transitionOut() { return cfgPassagesTransitionOut; },
			set transitionOut(value) {
				if (value != null) { // nullish test
					const valueType = getTypeOf(value);

					if (
						valueType !== 'string'
						&& (valueType !== 'number' || !Number.isSafeInteger(value) || value < 0)
					) {
						throw new TypeError(`Config.passages.transitionOut must be a string, non-negative integer number, null, or undefined (received: ${valueType})`);
					}
				}

				cfgPassagesTransitionOut = value;
			}

			/* [DEPRECATED] */
			/* eslint-disable comma-style */
			// Die if the deprecated passages descriptions getter is accessed.
			, get descriptions() {
				warnDeprecated(
					'Config.passages.descriptions',
					'Config.saves.descriptions',
					true
				);
			}
			// Warn if deprecated passages descriptions setter is assigned to,
			// then pass the value to the `Config.saves.descriptions` for
			// compatibilities sake.
			, set descriptions(value) {
				warnDeprecated(
					'Config.passages.descriptions',
					'Config.saves.descriptions'
				);

				switch (typeof value) {
					case 'boolean': {
						if (value && !Config.saves.descriptions) {
							Config.saves.descriptions = function () {
								return State.passage;
							};
						}

						break;
					}

					case 'function': {
						if (!Config.saves.descriptions) {
							Config.saves.descriptions = value;
						}

						break;
					}

					case 'undefined':
					case 'object': {
						if (value && !Config.saves.descriptions) {
							const dict = value;
							Config.saves.descriptions = function () {
								return Object.hasOwn(dict, State.passage) && dict[State.passage];
							};
						}

						break;
					}

					default:
						throw new TypeError(`Config.passages.descriptions must be a boolean, object, function, null, or undefined (received: ${getTypeOf(value)})`);
				}
			}
			/* eslint-enable comma-style */
			/* [/DEPRECATED] */
		}),

		/*
			Saves settings.
		*/
		saves : Object.freeze({
			get descriptions() { return cfgSavesDescriptions; },
			set descriptions(value) {
				if (!(value == null || value instanceof Function)) { // nullish test
					throw new TypeError(`Config.saves.descriptions must be a function, null, or undefined (received: ${getTypeOf(value)})`);
				}

				cfgSavesDescriptions = value;
			},

			get id() { return cfgSavesId; },
			set id(value) {
				if (typeof value !== 'string' || value === '') {
					throw new TypeError(`Config.saves.id must be a non-empty string (received: ${getTypeOf(value)})`);
				}

				cfgSavesId = value;
			},

			get isAllowed() { return cfgSavesIsAllowed; },
			set isAllowed(value) {
				if (!(value == null || value instanceof Function)) { // nullish test
					throw new TypeError(`Config.saves.isAllowed must be a function, null, or undefined (received: ${getTypeOf(value)})`);
				}

				cfgSavesIsAllowed = value;
			},

			get isEvalEnabled() { return cfgSavesEvalEnabled; },
			set isEvalEnabled(value) { 
				if (typeof value !== 'boolean') {
					throw new TypeError('Config.saves.isEvalEnabled must be a boolean');
				}
				cfgSavesEvalEnabled = value;
			},

			get maxAutoSaves() { return cfgSavesMaxAuto; },
			set maxAutoSaves(value) {
				if (!Number.isSafeInteger(value)) {
					throw new TypeError('Config.saves.maxAutoSaves must be an integer number');
				}
				else if (value < 0 || value > Save.MAX_INDEX + 1) {
					throw new RangeError(`Config.saves.maxAutoSaves out-of-bounds (range: 0–${Save.MAX_INDEX + 1}; received: ${value})`);
				}

				cfgSavesMaxAuto = value;
			},

			get maxSlotSaves() { return cfgSavesMaxSlot; },
			set maxSlotSaves(value) {
				if (!Number.isSafeInteger(value)) {
					throw new TypeError('Config.saves.maxSlotSaves must be an integer number');
				}
				else if (value < 0 || value > Save.MAX_INDEX + 1) {
					throw new RangeError(`Config.saves.maxSlotSaves out-of-bounds (range: 0–${Save.MAX_INDEX + 1}; received: ${value})`);
				}

				cfgSavesMaxSlot = value;
			},

			get metadata() { return cfgSavesMetadata; },
			set metadata(value) {
				if (!(value == null || value instanceof Function)) { // nullish test
					throw new TypeError(`Config.saves.metadata must be a function, null, or undefined (received: ${getTypeOf(value)})`);
				}

				cfgSavesMetadata = value;
			},

			get version() { return cfgSavesVersion; },
			set version(value) { cfgSavesVersion = value; }

			/* [DEPRECATED] */
			/* eslint-disable comma-style */
			// Internal only getter.
			, get _internal_autoload_() { // eslint-disable-line camelcase
				return cfgSavesAutoload;
			}
			// Warn if the deprecated autoload getter is accessed.
			, get autoload() {
				warnDeprecated(
					'Config.saves.autoload',
					'the Save.browser.continue API'
				);
				return cfgSavesAutoload;
			}
			// Warn if the deprecated autoload setter is assigned to.
			, set autoload(value) {
				warnDeprecated(
					'Config.saves.autoload',
					'the Save.browser.continue API'
				);

				if (value != null) { // nullish test
					const valueType = getTypeOf(value);

					if (
						valueType !== 'boolean'
						&& (valueType !== 'string' || value !== 'prompt')
						&& valueType !== 'function'
					) {
						throw new TypeError(`Config.saves.autoload must be a boolean, string ('prompt'), function, null, or undefined (received: ${valueType})`);
					}
				}

				cfgSavesAutoload = value;
			}

			// Die if the deprecated saves autosave getter is accessed.
			, get autosave() {
				warnDeprecated(
					'Config.saves.autosave',
					'Config.saves.maxAutoSaves and Config.saves.isAllowed',
					true
				);
			}
			// Die or warn if the deprecated saves autosave setter is assigned to,
			// while also setting `Config.saves.maxAutoSaves` and, possibly,
			// `Config.saves.isAllowed` for compatibilities sake.
			, set autosave(value) {
				switch (typeof value) {
					case 'boolean':
						warnDeprecated(
							'Config.saves.autosave',
							'Config.saves.maxAutoSaves'
						);
						break;

					case 'function': {
						warnDeprecated(
							'Config.saves.autosave',
							'Config.saves.isAllowed'
						);

						if (!Config.saves.isAllowed) {
							const callback = value;
							Config.saves.isAllowed = function (saveType) {
								// Allow all other types while testing auto saves.
								return saveType !== Save.Type.Auto || callback(saveType);
							};
						}

						break;
					}

					default: {
						warnDeprecated(
							'Config.saves.autosave',
							'Config.saves.isAllowed'
						);

						if (
							!(value instanceof Array)
							|| value.length === 0
							|| value.some(tag => typeof tag !== 'string')
						) {
							const valueType = getTypeOf(value);
							throw new TypeError(`Config.saves.autosave must be a boolean, Array<string>, function, null, or undefined (received: ${valueType}${valueType === 'Array' ? '<any>' : ''})`);
						}

						if (!Config.saves.isAllowed) {
							const userTags = value;
							Config.saves.isAllowed = function (saveType) {
								// Allow all other types while testing auto saves.
								return (
									saveType !== Save.Type.Auto
									|| userTags.includesAny(Story.get(State.passage).tags)
								);
							};
						}

						break;
					}
				}

				if (Config.saves.maxAutoSaves === 0) {
					Config.saves.maxAutoSaves = 1;
				}
			}

			// Die if the deprecated saves onLoad handler getter is accessed.
			, get onLoad() {
				warnDeprecated(
					'Config.saves.onLoad',
					'the Save.onLoad API',
					true
				);
			}
			// Warn if the deprecated saves onLoad handler setter is assigned to, then
			// pass the handler to the `Save.onLoad` API for compatibilities sake.
			, set onLoad(value) {
				warnDeprecated(
					'Config.saves.onLoad',
					'the Save.onLoad API'
				);
				Save.onLoad.add(value);
			}

			// Die if the deprecated saves onSave handler getter is accessed.
			, get onSave() {
				warnDeprecated(
					'Config.saves.onSave',
					'the Save.onSave API',
					true
				);
			}
			// Warn if the deprecated saves onSave handler setter is assigned to, then
			// pass the handler to the `Save.onSave` API for compatibilities sake.
			, set onSave(value) {
				warnDeprecated(
					'Config.saves.onSave',
					'the Save.onSave API'
				);
				Save.onSave.add(value);
			}

			// Die if the deprecated saves slots getter is accessed.
			, get slots() {
				warnDeprecated(
					'Config.saves.slots',
					'Config.saves.maxSlotSaves',
					true
				);
			}
			// Warn if the deprecated saves slots setter is assigned to, then pass
			// the value to the `Config.saves.maxSlotSaves` for compatibilities
			// sake.
			, set slots(value) {
				warnDeprecated(
					'Config.saves.slots',
					'Config.saves.maxSlotSaves'
				);
				Config.saves.maxSlotSaves = value;
			}

			// Warn if the deprecated saves tryDiskOnMobile getter is accessed, then
			// return `true`.
			, get tryDiskOnMobile() {
				warnDeprecated(
					'Config.saves.tryDiskOnMobile'
				);
				return true;
			}
			// Warn if the deprecated saves tryDiskOnMobile setter is assigned to.
			, set tryDiskOnMobile(value) {
				warnDeprecated(
					'Config.saves.tryDiskOnMobile'
				);
			}
			/* eslint-enable comma-style */
			/* [/DEPRECATED] */
		}),

		/*
			UI settings.
		*/
		ui : Object.freeze({
			get stowBarInitially() { return cfgUiStowBarInitially; },
			set stowBarInitially(value) {
				const valueType = getTypeOf(value);

				if (
					valueType !== 'boolean'
					&& (valueType !== 'number' || !Number.isSafeInteger(value) || value < 0)
				) {
					throw new TypeError(`Config.ui.stowBarInitially must be a boolean or non-negative integer number (received: ${valueType})`);
				}

				cfgUiStowBarInitially = value;
			},

			get updateStoryElements() { return cfgUiUpdateStoryElements; },
			set updateStoryElements(value) { cfgUiUpdateStoryElements = Boolean(value); }
		})
	});
})();
